mbs_results/ratio_of_means.py

```diff
@@ -0,0 +1,388 @@
+        ),
+        dict(
+            **default_columns,
+            **{"forward_or_backward": "b", "imputation_link": "b_link_question"}
+        ),
+    )
+
+    for args in cum_links_arguments:
+
+        df = get_cumulative_links(df, **args)
+
+    return df
+
+
+def count_impute_matches(
+    df: pd.DataFrame, **default_columns: Dict[str, str]
+) -> pd.DataFrame:
+    """Wrapper function to get counts of flagged matched pairs.
+
+    Parameters
+    ----------
+    df : pd.DataFrame
+        Original dataframe.
+    **default_columns : Dict[str, str]
+        The column names which were passed to ratio of means function.
+    Returns
+    -------
+    df : pd.DataFrame
+        Original dataframe with 3 new numeric columns.
+    """
+
+    count_arguments = [
+        dict(**default_columns, **{"flag": "f_match"}),
+        dict(**default_columns, **{"flag": "b_match"}),
+        dict(**default_columns, **{"flag": "flag_construction_matches"}),
+    ]
+
+    # TODO: count_matches return type not very convenient to combine all counts
+    # TODO: if count_matches returns a series then easier to combine them
+
+    all_counts = pd.DataFrame(columns=default_columns.values())
+
+    for args in count_arguments:
+        counts = count_matches(df, **args)
+        all_counts = counts.merge(all_counts, how="left")
+
+    return all_counts
+
+
+def ratio_of_means(
+    df: pd.DataFrame,
+    target: str,
+    period: str,
+    reference: str,
+    strata: str,
+    auxiliary: str,
+    filters: pd.DataFrame = None,
+) -> pd.DataFrame:
+    """
+    Imputes for each non-responding contributor a single numeric target
+    variable within the dataset for multiple periods simultaneously. It uses
+    the relationship between the target variable of interest and a predictive
+    value and/or auxiliary variable to inform the imputed value. The method
+    can apply forward, backward, construction or forward from construction
+    imputation. The type of imputation used will vary for each non-respondent
+    in each period depending on whether data is available in the predictive
+    period
+
+    Parameters
+    ----------
+    df : pd.DataFrame
+         Original dataframe.
+    target : str
+        Column name of values to be imputed.
+    period : str
+        Column name containing datetime information.
+    reference : str
+        Column name of unique Identifier.
+    strata : str
+        Column name containing strata information (sic).
+    auxiliary : str
+        Column name containing auxiliary information (sic).
+
+    Returns
+    -------
+    pd.DataFrame
+        Original dataframe with imputed values in the target column, and with
+        intermediate columns which were used for the imputation method.
+    """
+
+    # Saving args to dict, so we can pass same attributes to multiple functions
+    # These arguments are used from the majority of functions
+    # Remove df to allow pipe chain
+
+    default_columns = locals()
```

Comment:
Good point, we need to chat about the `config` format and update the wrapper accordingly

Reply:
Added kwargs : mapping, optional in the function, this will allow passing a config file as a dictionary. This is untested because in json file period is set to period , while test cases have it as date  , also no auxiliary is defined in config.